### ISSUE TITLE
Confusion about the loss function

### ISSUE THREAD

**Issue by SenYan1999** [nianlonggu/Local-Citation-Recommendation#1]
Hi Nianlong, 

I am a little confused about the Triplet Loss Function part. I know that sim_mat's dimension is batch_size * batch_size, but why mask's dim is batch_size * batch_size * batch_size? For example, what does mask[1, 2, 3] mean? My guess is that mask[1, 2, 3] means sim(paper 1, paper 3) - sim(paper 1, paper 2) and mask [1, 3, 2] means sim(paper 1, paper 2) - sim(paper 1, paper 3).

So your core idea here is to build all the combination of sim(paper a, paper b) - sim(paper a, paper c) and use mask to select the valid pairs. Is it correct? 

Thanks for your attention! 

```python
def forward( self, feats, class_labels, irrelevance_levels, positive_irrelevance_levels = [1], similarity = "cosine" ):

    sim_mat = self.pairwise_similarity( feats, similarity )
    mask = self.get_mask(  class_labels, irrelevance_levels , positive_irrelevance_levels = positive_irrelevance_levels )
    margin = self.get_margin(  class_labels, irrelevance_levels  )

    sim_difference = sim_mat.unsqueeze(1) - sim_mat.unsqueeze(2) + margin
    sim_difference = sim_difference * mask
    loss = torch.clamp(sim_difference, min = 0 ).sum()

    return loss
```

**Comment by nianlonggu** [comment]
> Hi Nianlong,
> 
> I am a little confused about the Triplet Loss Function part. I know that sim_mat's dimension is batch_size * batch_size, but why mask's dim is batch_size * batch_size * batch_size? For example, what does mask[1, 2, 3] mean? My guess is that mask[1, 2, 3] means sim(paper 1, paper 3) - sim(paper 1, paper 2) and mask [1, 3, 2] means sim(paper 1, paper 2) - sim(paper 1, paper 3).
> 
> So your core idea here is to build all the combination of sim(paper a, paper b) - sim(paper a, paper c) and use mask to select the valid pairs. Is it correct?
> 
> Thanks for your attention!
> 
> ```python
> def forward( self, feats, class_labels, irrelevance_levels, positive_irrelevance_levels = [1], similarity = "cosine" ):
> 
>     sim_mat = self.pairwise_similarity( feats, similarity )
>     mask = self.get_mask(  class_labels, irrelevance_levels , positive_irrelevance_levels = positive_irrelevance_levels )
>     margin = self.get_margin(  class_labels, irrelevance_levels  )
> 
>     sim_difference = sim_mat.unsqueeze(1) - sim_mat.unsqueeze(2) + margin
>     sim_difference = sim_difference * mask
>     loss = torch.clamp(sim_difference, min = 0 ).sum()
> 
>     return loss
> ```

The sim_mat is batchsize x batchsize, but the sim_difference is batchsize x batchsize x batchsize, because there is a unsqueeze func when computing sim_difference. Your understanding is correct. All the invalid triplets will masked off. Meanwhile, the relevance level just provide a convenient way to determine which one is invalid - the positive document should be more relevant to the achor doc than the negative document, otherwise it is invalid.

**Comment by SenYan1999** [comment]
Thanks, your implementation is very smart and effective! Appreciate for your work!
